### PR TITLE
let expression support

### PR DESCRIPTION
**Pull request by Dspil**
Syntax:

`let <short variable declaration> in <expression>`

The short variable declaration allows for multiple let bound variables in a single let expression:

`let x,y := z+1,w+2 in x - y`

Expressions assigned to `let` variables should be pure and the result of the let expression should also be pure.

### CODE REVIEW

**Comment by ArquintL** [comment]
Finally 🥳
Thanks a lot!

File: src/main/scala/viper/gobra/translator/encodings/typeless/AssertionEncoding.scala [hunk 1]
```diff
@@ -54,6 +54,13 @@ class AssertionEncoding extends Encoding {
         case Seq() => newExists
         case errors => Violation.violation(s"invalid trigger pattern (${errors.head.readableMessage})")
       }
+
+    case let: in.Let =>
+      for {
+        exp <- ctx.expression(let.in)
+        l = ctx.variable(let.left)
+        r <- ctx.expression(let.right)
+      } yield withSrc(vpr.Let(l, r, exp), let)
```
Comment (Felalolf) [inline src/main/scala/viper/gobra/translator/encodings/typeless/AssertionEncoding.scala:63]: The writer also has features to introduce let bindings, for instance `_ <- bind(l, exp)`, but this encoding should also be fine.

**Review by Felalolf — COMMENTED** [review]
One mistake in the ast definition.

File: src/main/scala/viper/gobra/frontend/info/implementation/typing/ghost/separation/GhostWellDef.scala [hunk 2]
```diff
@@ -114,6 +114,7 @@ trait GhostWellDef { this: TypeInfoImpl =>
        | _: PBitNegation
        | _: PBinaryExp[_,_]
        | _: PUnfolding
+       | _: PLet
```
Comment (Felalolf) [inline src/main/scala/viper/gobra/frontend/info/implementation/typing/ghost/separation/GhostWellDef.scala:117]:
```suggestion
```
This is now redundant

File: src/main/scala/viper/gobra/frontend/info/implementation/typing/ghost/GhostExprTyping.scala [hunk 3]
```diff
@@ -410,6 +415,7 @@ trait GhostExprTyping extends BaseTyping { this: TypeInfoImpl =>
       })
 
       case _: PUnfolding => true
+      case _: PLet => true
```
Comment (Felalolf) [inline src/main/scala/viper/gobra/frontend/info/implementation/typing/ghost/GhostExprTyping.scala:418]:
```suggestion
      case _: PLet => true // the well-definedness check makes sure that both sub-expressions are pure.
```

**Review by Felalolf — APPROVED** [review]
Two more small changes.

**Pull request merged**